File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
    let level = 0;
    const mp = new Map<number,number[]>();
    let ans = new Array<number[]>();

    const traverse = (root : TreeNode | null) => {
        if(root === null) return;
        if(mp.has(level)) mp.get(level).push(root.val)
        else mp.set(level , [root.val]);
        level ++;
        traverse(root.left);
        traverse(root.right);
        level --;
    }
    traverse(root);
    // console.log(mp);

    for(let [level,value] of mp){
        ans.push(value);
    }

    return ans;
};

File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null) return ans;        

        Queue<TreeNode> q = new ArrayDeque<>();
        q.offer(root);

        while(!q.isEmpty()){
            int curQueueSize = q.size();
            List<Integer> curAns = new ArrayList<>();
            for(int i = 0; i <curQueueSize; i ++) {
                TreeNode curElm = q.poll();
                if(curElm == null) continue;
                if(curElm.left != null) q.offer(curElm.left);
                if(curElm.right != null) q.offer(curElm.right);
                curAns.add(curElm.val);
            }
            ans.add(curAns);
        }
        return ans;
    }
}